refact(diffgeom) : redesign CoordSystem

JSS95 opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed

`Manifold`, `Patch` and `CoordSystem` are now just `Basic`, and not `Atom` anymore. An object being a leaf node should be indicated by empty `.args`, not being a subclass of certain class. (See #21017)

`CoordSystem` now uses simpler construction signature. Previously, transformation relation was built using `Lambda`. Now, only a tuple is suffice.


#### Other comments


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
NO ENTRY
<!-- END RELEASE NOTES -->

sylee957 commented on sympy/diffgeom/diffgeom.py:
```diff
@@ -499,17 +572,71 @@ def coord_tuple_transform_to(self, to_sys, coords):
             result = result.subs(list(zip(self.symbols, coordinates)))
         return result
-    jacobian_matrix = jacobian
+    jacobian_mat = jacobian_matrix = jacobian
```
Is it necessary to have these shortcuts?

JSS95 replied:
probably not

sylee957 commented on sympy/diffgeom/diffgeom.py:
```diff
@@ -320,11 +349,8 @@ def manifold(self):
-            ) for i,s in enumerate(self.args[2])
-        ]
+        return Tuple(*[CoordinateSymbol(self, i, **s._assumptions.generator)
+            for i,s in enumerate(self.args[2])])
```
I would consider using built-in `tuple` than `Tuple` if this is not necessary.

sylee957 commented on sympy/diffgeom/diffgeom.py:
```diff
@@ -499,17 +572,70 @@ def coord_tuple_transform_to(self, to_sys, coords):
         if coordinates is not None:
             result = result.subs(list(zip(self.symbols, coordinates)))
         return result
-    jacobian_matrix = jacobian
```
No, I don't think that you should remove this line completely because it was there in the previous version.
But I would rather remove `jacobian_det` below if that is also accidental

JSS95 replied:
Oops. This was a mistake.

BTW I think the real names for two methods should be `jacobian_mat()` and `jacobian_det()` for clarity, since the term "jacobian" is ambiguous. (https://mathworld.wolfram.com/Jacobian.html)

JSS95 replied:
I don't want this topic to be a blocker for this PR so I'll open a separate issue for the naming of methods.

JSS95 committed:
refact(diffgeom) : redesign CoordSystem

Make Manifold Basic, not Atom
Make Patch Basic, not Atom
Make CoordSystem Basic, not Atom
Make signature of CoordSystem simpler

JSS95 committed:
refact(diffgeom) : redesign CoordSystem

Fix srepr printing of CoordinateSymbol
Enhance doc
